util/chunk/chunk_allocator.go

```diff
@@ -0,0 +1,247 @@
+package chunk
+
+import (
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/types"
+)
+
+var (
+	// GlobalAllocator is the root allocator of all allocators.
+	GlobalAllocator = NewMultiBufAllocator(uint(runtime.NumCPU()), 15, 64)
+)
+
+// Allocator is used to manage byte slices.
+type Allocator interface {
+	Alloc(l, c int) []byte
+	Free(buf []byte)
+	SetParent(a Allocator)
+	Close()
+}
+
+type multiBufChan struct {
+	allocIndex    uint32
+	freeIndex     uint32
+	numAllocators uint32
+	maxCap        int
+	allocators    []Allocator
+}
+
+// NewMultiBufAllocator creates a multiBufChan.
+func NewMultiBufAllocator(numAllocators, bitN, bufSize uint) Allocator {
+	m := &multiBufChan{0, 0, uint32(numAllocators), 1 << bitN, make([]Allocator, 0, numAllocators)}
+	for i := 0; i < int(numAllocators); i++ {
+		m.allocators = append(m.allocators, NewBufAllocator(bitN, bufSize))
+	}
+	return m
+}
+
+// Alloc alloc memory.
+func (b *multiBufChan) Alloc(l, c int) []byte {
+	return b.allocators[atomic.AddUint32(&b.allocIndex, 1)%b.numAllocators].Alloc(l, c)
+}
+
+// Free releases memory.
+func (b *multiBufChan) Free(buf []byte) {
+	b.allocators[atomic.AddUint32(&b.freeIndex, 1)%b.numAllocators].Free(buf)
+}
+
+// SetParent sets a parent for this allocator.
+func (b *multiBufChan) SetParent(a Allocator) {
+	for _, x := range b.allocators {
+		x.SetParent(a)
+	}
+}
+
+// Close closes this allocator.
+func (b *multiBufChan) Close() {
+	for _, x := range b.allocators {
+		x.Close()
+	}
+}
+
+var (
+	capIndexBit           = 15
+	allocIndex, freeIndex = getCapIndex(uint(capIndexBit))
+	pad                   = make([]byte, (1<<uint(capIndexBit))+1)
+)
+
+func getCapIndex(bitN uint) ([]int, []int) {
+	allocIndex := make([]int, int(1<<bitN)+1)
+	freeIndex := make([]int, int(1<<bitN)+1)
+	for i := uint(1); i <= bitN; i++ {
+		left := 1 << (i - 1)
+		right := 1 << i
+		for j := left + 1; j <= right; j++ {
+			allocIndex[j] = int(i)
+		}
+		for j := left; j < right; j++ {
+			freeIndex[j] = int(i - 1)
+		}
+	}
+	allocIndex[1] = 0
+	freeIndex[1<<bitN] = int(bitN)
+	return allocIndex, freeIndex
+}
+
+type bufChan struct {
+	maxCap     int
+	bufList    []chan []byte
```

Comment:
This struct create too many small objects and bad for GC

Reply:
You are right and I will fix this problem soon.